File: androidApp/src/main/java/com/example/kmmnews/android/NewsViewModel.kt
package com.example.kmmnews.android

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.kmmnews.NewsArticle
import com.example.shared.NewsRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class NewsViewModel(private val newsRepository: NewsRepository) : ViewModel() {

    // StateFlow for news articles
    private val _newsArticles = MutableStateFlow<List<NewsArticle>>(emptyList())
    val newsArticles: StateFlow<List<NewsArticle>> = _newsArticles

    // StateFlow for selected news article
    private val _selectedNewsArticle = MutableStateFlow<NewsArticle?>(null)
    val selectedNewsArticle: StateFlow<NewsArticle?> = _selectedNewsArticle

    // StateFlow for loading state
    private val _loading = MutableStateFlow(false)
    val loading: StateFlow<Boolean> = _loading

    // Function to fetch news articles
    fun fetchNews(query: String, fromDate: String) {
        viewModelScope.launch {
            _loading.value = true // Set loading to true when fetching starts
            try {
                _newsArticles.value = newsRepository.fetchNews(query, fromDate) ?: emptyList()
            } finally {
                _loading.value = false // Set loading to false when fetching ends
            }
        }
    }

    // Function to fetch details of a specific news article
    fun fetchNewsDetails(articleUrl: String) {
        _selectedNewsArticle.value = _newsArticles.value.find { it.url == articleUrl }
    }
}

File: iosApp/iosApp/NewsViewModel.swift
import SwiftUI
import shared

class NewsViewModel: ObservableObject {
    // Published properties to update the UI
    @Published var articles: [NewsArticle] = []
    @Published var selectedArticle: NewsArticle?
    @Published var isLoading: Bool = false

    private let repository: NewsRepository

    // Initialize with the shared repository
    init(repository: NewsRepository) {
        self.repository = repository
        fetchNews() // Optionally fetch news when the view model is initialized
    }

    // Function to fetch news articles
    func fetchNews(query: String = "technology", fromDate: String = "2024-07-12") {
        isLoading = true
        repository.fetchNews(query: query, fromDate: fromDate) { [weak self] news, error in
            DispatchQueue.main.async {
                if let news = news {
                    self?.articles = news
                } else if let error = error {
                    print("Error fetching news: \(error)")
                    self?.articles = []
                }
                self?.isLoading = false
            }
        }
    }

    // Function to fetch the details of a specific news article
    func fetchNewsDetails(articleUrl: String) {
        selectedArticle = articles.first { $0.url == articleUrl }
    }
}

